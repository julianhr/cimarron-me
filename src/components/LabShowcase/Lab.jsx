import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { css } from '@emotion/core'

import ImageLoader from '../library/ImageLoader'
import AnchorTag from '../library/AnchorTag'


const baseArticle = css`
  border-radius: 5px;
  display: flex;
  flex-direction: column;
`

const STYLES = {
  link: css`
    ${baseArticle}
    max-width: 250px;
    text-decoration: none;
    color: unset;
    background: #ededed;
    padding: 5px 12px 12px;
    will-change: box-shadow;
    transition: box-shadow 30ms ease-in;

    :hover {
      box-shadow: 1px 1px 4px #bbb;
    }
  `
}

const Article = styled.article`
  ${baseArticle}
  padding: 10px 15px;
  width: 100%;
  height: 220px;
  align-items: center;
`

const TechRoot = styled.section`
  display: flex;
  flex-wrap: wrap;
  font-size: 13px;
`

const Tech = styled.div`
  margin: 2px;
  padding: 2px 7px;
`

function Lab({ title, stack, urlPath, thumbnailSrc, isLinkRouted }) {
  const getTechStyle = (index) => {
    const hue = (330 + index * 30) % 360
    const sat = 60
    const lum = 85
    return { background: `hsl(${hue}, ${sat}%, ${lum}%)` }
  }

  const renderTech = stack => (
    stack.map((tech, i) => (
      <Tech
        key={i}
        css={getTechStyle(i)}
      >
        {tech}
      </Tech>
    ))
  )

  return (
    <AnchorTag
      url={urlPath}
      rootStyle={STYLES.link}
      isLinkRouted={isLinkRouted}
    >
      <Article>
        <ImageLoader
          imgSrc={thumbnailSrc}
          maxWidth={200}
          maxHeight={120}
          styles={{ root: { marginBottom: 20 } }}
        />
        <h3>{title}</h3>
      </Article>
      <TechRoot>
        {renderTech(stack)}
      </TechRoot>
    </AnchorTag>
  )
}

Lab.propTypes = {
  thumbnailSrc: PropTypes.string.isRequired,
  stack: PropTypes.arrayOf(PropTypes.string).isRequired,
  title: PropTypes.string.isRequired,
  urlPath: PropTypes.string.isRequired,
  isLinkRouted: PropTypes.bool.isRequired,
}

export default Lab
