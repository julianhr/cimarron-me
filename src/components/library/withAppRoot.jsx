import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from 'emotion-theming'
import { Global } from '@emotion/core'
import 'sanitize.css'

import appStore from '~/reducers'
import globalStyles from '~/styles/globalStyles'
import theme from '~/styles/theme'


function withAppRoot(WrappedComponent, otherProps={}) {
  return class extends React.Component {
    render() {
      return (
        <Provider store={appStore}>
          <BrowserRouter>
            <ThemeProvider theme={theme}>
              <Global styles={globalStyles} />
              <WrappedComponent {...this.props} {...otherProps} />
            </ThemeProvider>
          </BrowserRouter>
        </Provider>
      )
    }
  }
}

export default withAppRoot
