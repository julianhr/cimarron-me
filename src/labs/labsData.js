import React from 'react'


export const LAB_SHAPE = {
  title: ['string'],
  urlPath: ['string'],
  thumbnailSrc: ['string'],
  stack: ['object'],
  createdAt: ['string']
}

const DATA = [
  {
    title: 'Infinite Scroller',
    urlPath: '/labs/infinite-scroller',
    thumbnailSrc: require('../labs/001_InfiniteScroller/thumbnail.png'),
    stack: 'React Flask AWS'.split(' '),
    createdAt: '2019-03',
  },
  {
    title: 'Spreadsheet App',
    urlPath: 'https://julianhr.github.io/app-spreadsheet',
    thumbnailSrc: require('../apps/001_Spreadsheet/thumbnail.png'),
    stack: 'React xstate'.split(' '),
    createdAt: '2019-07',
  },
]

export default DATA
