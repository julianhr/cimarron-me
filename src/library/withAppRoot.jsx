import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from 'emotion-theming'
import { Global } from '@emotion/core'
import 'sanitize.css'

import globalStyles from '~/styles/globalStyles'
import theme from '~/styles/theme'


function withAppRoot(WrappedComponent, store) {
  return class extends React.Component {
    render() {
      return (
        <Provider store={store}>
          <BrowserRouter>
            <ThemeProvider theme={theme}>
              <Global styles={globalStyles} />
              <WrappedComponent {...this.props} />
            </ThemeProvider>
          </BrowserRouter>
        </Provider>
      )
    }
  }
}

export default withAppRoot
