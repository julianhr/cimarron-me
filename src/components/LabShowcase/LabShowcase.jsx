import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import Lab from './Lab'


const Root = styled.section`
  display: flex;
  flex-wrap: wrap;
  padding: 80px 20px 20px;

  @media (max-width: 540px) {
    justify-content: center;
  }
`

function LabShowcase({ labs }) {
  const renderLabs = labs.map((lab, i) => (
    <Lab
      key={i}
      isLinkRouted={false}
      {...lab}
    />
  ))

  return (
    <Root>
      {renderLabs}
    </Root>
  )
}

LabShowcase.propTypes = {
  labs: PropTypes.arrayOf(PropTypes.object).isRequired,
}

export default LabShowcase
