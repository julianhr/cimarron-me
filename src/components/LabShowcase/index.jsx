import LabShowcase from './LabShowcase'
export default LabShowcase
