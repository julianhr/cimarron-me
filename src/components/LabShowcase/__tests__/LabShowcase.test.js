import React from 'react'
import { render } from 'enzyme'

import LabShowcase from '../LabShowcase'
import withAppRoot from '~/components/library/withAppRoot'
import labsData from '~/__tests__/__fixtures__/labsDataFixture'


describe('LabShowcase', () => {
  it('should match snapshot', () => {
    const RootedLabShowcase = withAppRoot(LabShowcase, { labs: labsData })
    const wrapper = render(<RootedLabShowcase />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('required props', () => {
    it('labs', () => {
      const RootedLabShowcase = withAppRoot(LabShowcase)
      expect(() => render(<RootedLabShowcase />)).toThrow()
    })
  })
})
