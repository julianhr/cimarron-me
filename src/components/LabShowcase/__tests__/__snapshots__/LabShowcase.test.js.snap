// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabShowcase should match snapshot 1`] = `
<section
  class="css-1lzrryp-Root e1uula6j0"
>
  <a
    class="css-4yofyo-baseArticle-link-link"
    href="/path/1"
    rel="noopener noreferrer"
  >
    <article
      class="css-1dt8rsn-Article-baseArticle e1lgusa60"
    >
      <figure
        class="css-1t2crv9-Root-ImageLoader eeswxza0"
      >
        <img
          class="css-1k90mm3"
          src="http://image1.com"
        />
      </figure>
      <h3
        class="css-1brvuev-H3 e1lgusa63"
      >
        Title 1
      </h3>
    </article>
    <section
      class="css-1k4ntzt-TechRoot e1lgusa61"
    >
      <div
        class="css-if30yk-Tech e1lgusa62"
      >
        React
      </div>
      <div
        class="css-1h4zuz1-Tech e1lgusa62"
      >
        Python
      </div>
    </section>
  </a>
  <a
    class="css-4yofyo-baseArticle-link-link"
    href="/path/2"
    rel="noopener noreferrer"
  >
    <article
      class="css-1dt8rsn-Article-baseArticle e1lgusa60"
    >
      <figure
        class="css-1t2crv9-Root-ImageLoader eeswxza0"
      >
        <img
          class="css-1k90mm3"
          src="http://image2.com"
        />
      </figure>
      <h3
        class="css-1brvuev-H3 e1lgusa63"
      >
        Title 2
      </h3>
    </article>
    <section
      class="css-1k4ntzt-TechRoot e1lgusa61"
    >
      <div
        class="css-if30yk-Tech e1lgusa62"
      >
        Ruby
      </div>
      <div
        class="css-1h4zuz1-Tech e1lgusa62"
      >
        Go
      </div>
    </section>
  </a>
</section>
`;
